File: packages/python/python-shared/src/python_shared/http/pipeline/step_circuit_breaker.py
from python_shared.http.constants import HTTP_CONSTANTS
from python_shared.http.pipeline.types import PipelineContext

class StepCircuitBreaker:
    name: str = "CircuitBreaker"
    description: str = "Bounded LRU Circuit Breaker State Verification"

    async def execute(self, ctx: PipelineContext) -> None:
        ctx.step_index += 1
        ctx.route_template = ctx.config.url
        ctx.circuit_key = ctx.circuit_breaker.get_circuit_key(ctx.tenant_id, ctx.route_template)

        if not ctx.circuit_breaker.can_execute(ctx.circuit_key):
            state = ctx.circuit_breaker.get_state(ctx.circuit_key)
            state_name = state.state if state else HTTP_CONSTANTS.CIRCUIT_OPEN
            raise RuntimeError(f"Circuit breaker is {state_name} for key {ctx.circuit_key}")

File: packages/node/shared-infra/src/http/pipeline/step-circuit-breaker.ts
/**
 * @file step-circuit-breaker.ts
 * @description Pipeline Step 7: Bounded Circuit Breaker State Verification.
 */

import type { PipelineStep, PipelineContext } from "./types";
import { deriveRouteTemplate } from "../utils/http-utils";
import { HTTP_CONSTANTS } from "../constants";

export class StepCircuitBreaker implements PipelineStep {
  public readonly name = "CircuitBreaker";
  public readonly description = "Bounded LRU Circuit Breaker State Verification";

  public async execute<T>(ctx: PipelineContext<T>): Promise<void> {
    ctx.stepIndex++;
    ctx.routeTemplate = deriveRouteTemplate(ctx.config.url);
    ctx.circuitKey = ctx.circuitBreaker.getCircuitKey(ctx.tenantId, ctx.routeTemplate);

    if (!ctx.circuitBreaker.canExecute(ctx.circuitKey)) {
      const state = ctx.circuitBreaker.getState(ctx.circuitKey);
      ctx.span?.setAttribute(HTTP_CONSTANTS.KEY_CIRCUIT_STATE, state?.state || HTTP_CONSTANTS.CIRCUIT_OPEN);
      ctx.span?.addEvent(HTTP_CONSTANTS.EVENT_CIRCUIT_EVALUATED, {
        "circuit.state": state?.state,
        "circuit.key": ctx.circuitKey,
      });
      throw new Error(`Circuit breaker is ${state?.state || "OPEN"} for key ${ctx.circuitKey}`);
    }
  }
}
